### PR TITLE
sstables: add dead row count when issuing warning to system.large_partitions

### LINKED ISSUE #13968: Large data handler does not take range tombstones into account
For detecting large partitions with too many rows we count only rows, but we don't count range tombstone markers.
Since the range tombstones need to be processed and accumulation of them may reduce performance we should count them too.  Since they are associated with clustering positions it makes sense to use the same counter as part as total number of rows for the purpose of threshold detection and reporting, but to add a separate counter (and respective column in the large_partitions system table) that would count only range tombstone markers (that correspond with range_tombstone_change mutation fragments in memory).

With that in mind, it'd also make sense to distinguish between live and dead rows, and to add another counter, similar in logic to the one described above for range tombstones, that would count the number of dead rows out of the total number of rows (what that crosses the threshold).  This way when a partition is detected with "too many rows", we'd know how many of which are dead and how many are range tombstone markers (the rest are alive).

### PR DESCRIPTION
**Pull request by nodep**
This is the second half of the fix for issue #13968. The first half is already merged with PR #18346

Scylla issues warnings for partitions containing more rows than a configured threshold. The warning is issued by inserting a row into the `system.large_partitions` table. This row contains the information about the partition for which the warning is issued: keyspace, table, sstable, partition key and size, compaction time and the number of rows in the partition. A previous PR #18346 also added range tombstone count to this row.

This change adds a new counter for dead rows to the large_partitions table.

This change also adds cluster feature protection for writing into these new counters. This is needed in case a cluster is in the process of being upgraded to this new version, after which an upgraded node writes data with the new schema into `system.large_partitions`, and finally a node is then rolled back to an old version. This node will then revert the schema to the old version, but the written sstables will still contain data with the new counters, causing any readers of this table to throw errors when they encounter these cells.

This is an enhancement, and backporting is not needed.

Fixes #13968

### CODE REVIEW

File: gms/feature_service.hh [hunk 1]
```diff
@@ -124,6 +124,7 @@ public:
     gms::feature aggregate_storage_options { *this, "AGGREGATE_STORAGE_OPTIONS"sv };
     gms::feature collection_indexing { *this, "COLLECTION_INDEXING"sv };
     gms::feature large_collection_detection { *this, "LARGE_COLLECTION_DETECTION"sv };
+    gms::feature deleted_rows_detection { *this, "DELETED_ROWS_DETECTION"sv };
```
Comment (denesb) [inline gms/feature_service.hh:127]: This name is too generic, a better name would be `LARGE_PARTITION_DETECTOR_DEAD_ROWS_DETECTION`.

File: db/large_data_handler.cc [hunk 2]
```diff
@@ -120,12 +120,21 @@ cql_table_large_data_handler::cql_table_large_data_handler(gms::feature_service&
     , _record_large_cells([this] (const sstables::sstable& sst, const sstables::key& pk, const clustering_key_prefix* ck, const column_definition& cdef, uint64_t cell_size, uint64_t collection_elements) {
         return internal_record_large_cells(sst, pk, ck, cdef, cell_size, collection_elements);
     })
-    , _feat_listener(_feat.large_collection_detection.when_enabled([this] {
+    , _record_large_partitions([this] (const sstables::sstable& sst, const sstables::key& pk, uint64_t partition_size, uint64_t rows, uint64_t range_tombstones, uint64_t dead_rows) {
+        return internal_record_large_partitions(sst, pk, partition_size, rows);
+    })
```
Comment (denesb) [inline db/large_data_handler.cc:125]: A few sentences in the commitlog description, describing what is happening here, would have made the review easier.

File: sstables/mx/writer.cc [hunk 3]
```diff
@@ -1259,7 +1259,7 @@ void writer::write_clustered(const clustering_row& clustered_row, uint64_t prev_
         }
     }
 
-    const bool is_dead = !clustered_row.is_live(_schema, clustered_row.tomb().regular(), gc_clock::now());
+    const bool is_dead = !clustered_row.is_live(_schema, tombstone(), gc_clock::now());
```
Comment (denesb) [inline sstables/mx/writer.cc:1262]: Please fold this into the commit which introduced this line, AFAIK said commit is also introduced in this PR.

File: gms/feature_service.hh (hunk 1 above)
Comment (nodep) [reply gms/feature_service.hh]: This change adds cluster feature protection for two new counters in `system.large_partitions`: `dead_rows` and `range_tombstones`. So I have renamed the cluster feature to `RANGE_TOMBSTONE_AND_DEAD_ROWS_DETECTION`.

File: db/large_data_handler.cc (hunk 2 above)
Comment (nodep) [reply db/large_data_handler.cc:125]: I added new a description.

File: sstables/mx/writer.cc (hunk 3 above)
Comment (nodep) [reply sstables/mx/writer.cc:1262]: I have folded the coommits.

File: docs/dev/system_keyspace.md [hunk 4]
```diff
@@ -33,6 +33,7 @@ CREATE TABLE system.large_partitions (
     partition_size bigint,
     partition_key text,
     range_tombstones bigint,
+    dead_rows bigint,
```
Comment (bhalevy) [inline docs/dev/system_keyspace.md:36]: hmm, these two columns were added last, after `compaction_time` (they have to, for backward compatibility). So the doc should reflect that.

File: docs/troubleshooting/large-partition-table.rst [hunk 5]
```diff
@@ -111,9 +113,10 @@ Large partitions are stored in a system table with the following schema:
        sstable_name text,
        partition_size bigint,
        partition_key text,
-       rows bigint,
-       range_tombstones bigint,
        compaction_time timestamp,
+       dead_rows bigint,
+       range_tombstones bigint,
+       rows bigint,
```
Comment (bhalevy) [inline docs/troubleshooting/large-partition-table.rst:119]: here too, lets describe the actual schema.

File: docs/troubleshooting/large-partition-table.rst [hunk 6]
```diff
@@ -73,9 +75,9 @@ Example output:
 
 .. code-block:: console
 
-   keyspace_name | table_name | sstable_name     | partition_size | partition_key                                       | range_tombstones | rows | compaction_time
-   --------------+------------+------------------+----------------+-----------------------------------------------------+------------------+------+---------------------
-          demodb |       tmcr | md-6-big-Data.db |     1188716932 | {key: pk{000400000001}, token:-4069959284402364209} |                0 | 1942 | 2018-07-23 08:10:34
+   keyspace_name | table_name | sstable_name     | partition_size | partition_key                                       | dead_rows | range_tombstones | rows | compaction_time
```
Comment (bhalevy) [inline docs/troubleshooting/large-partition-table.rst:78]: please try the query in the example (SELECT *) on real data and make sure the example output reflects the same order of columns.

**Comment by bhalevy** [comment]
looks good overall.
I had some comments about the documentation.

File: docs/troubleshooting/large-partition-table.rst (hunk 5 above)
Comment (bhalevy) [reply docs/troubleshooting/large-partition-table.rst:119]: Hmm, I wonder if this isn't a bug with DESCRIBE TABLE.
It should produce a compatible schema definition, and that should have exactly the same order of columns.
Maybe I'm missing something.
@tgrabiec / @denesb any insights?

Comment (tgrabiec) [reply docs/troubleshooting/large-partition-table.rst:119]: When schema is persisted to schema tables, the order in which columns were passed to the schema builder is lost. Order of regular columns after reading from schema tables is lexicographical by column name. So "describe" output is correct.

File: docs/troubleshooting/large-partition-table.rst (hunk 6 above)
Comment (nodep) [reply docs/troubleshooting/large-partition-table.rst]: I have changed the order of the columns to reflect the order in which the `SELECT * FROM system.large_partitions;` lists the values.

File: docs/troubleshooting/large-partition-table.rst (hunk 5 above)
Comment (tgrabiec) [reply docs/troubleshooting/large-partition-table.rst:119]: Yes, column matching happens by name, not by index in the schema object. You can read old sstables using newer schema which has more columns.

Comment (bhalevy) [reply docs/troubleshooting/large-partition-table.rst:119]: For some reason I remembered we're relying on the column id, not its name, maybe for column rename...

File: docs/dev/system_keyspace.md (hunk 4 above)
Comment (bhalevy) [reply docs/dev/system_keyspace.md:36]: edit: schema sorts the columns by name, not by order of insertion to schema_builder.
Backward compatibility is achieved by matching column names, not column id's (for better or worse)

**Review by bhalevy — APPROVED** [review]

**Pull request merged**